guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,166 @@
+
+With the GVL, using a lot of threads has diminishing returns. A Rails app rarely benefits from more than 6. To have a large number of workers, some other concurrency method should be used.
+
+Similarly, Ruby's garbage collector is "stop-the-world" so when it's triggered all threads have to stop. This also means diminishing returns for large numbers of threads.
+
+Threads are less resilient than processes. Certain errors like segmentation faults can destroy the entire process and all threads inside. A single request allocating a lot of memory can stop all threads while the garbage collector runs.
+
+### Hybrid Concurrency
+
+Puma allows forking multiple processes, each of which uses multiple threads. This provides a compromise between process-based and thread-based concurrency. Using multiple threads per process helps memory usage. Multiple processes permit running more Ruby code at the same time since there is one GVL per process.
+
+Hybrid concurrency limits the damage from a segmentation fault or other error that kills a process. A single process dying will kill that process's threads but not threads in other processes. Each process has its own independent garbage collector.
+
+Choosing Default Settings
+-------------------------
+
+Rails' default settings aren't appropriate for all application sizes. You can improve performance for your I/O-heavy application that serves a lot of requests by changing settings.
+
+This section contains common sense defaults based on the type and size of your application and the hosts on which it runs. You can improve performance more by testing your application specifically. See "Performance Testing" for details.
+
+These are production recommendations. Your development application will have different needs from a production application, and should use a different configuration.
+
+[Puma's deployment documentation](https://github.com/puma/puma/blob/master/docs/deployment.md) may also be useful.
+
+### Threads Per Process
+
+Rails uses 3 threads per process by default. A well-optimized I/O-heavy Rails application should specify 5 or 6 threads per process at maximum. Discourse, for example, benefits from about 5 threads per process. Discourse also executes many database queries per request and frequently uses Redis. More self-contained applications with fewer database and API queries benefit from around 3 threads per process.
+
+The default Puma configuration mentions that "as a rule of thumb, increasing the number of threads will increase how much traffic a given process can handle (throughput), but due to CRuby's Global VM Lock (GVL) it has diminishing returns and will degrade the response time (latency) of the application."
+
+To set the number of threads, you can change the call to the +threads+ method in +config/puma.rb+. Or you can set the +RAILS_MAX_THREADS+ environment variable, which will do the same. Make sure your +config/database.yml+ file sets +pool+ to be at least as high as the number of threads.
```

Comment:
```suggestion
To set the number of threads, you can change the call to the `threads` method in `config/puma.rb`. Or you can set the `RAILS_MAX_THREADS` environment variable, which will do the same. Make sure your `config/database.yml` file sets `pool` to be at least as high as the number of threads.
```

I'm pretty sure we have to use markdown ticks here for `<tt>` formatting.

Reply:
Maybe? I don't normally use the plus here, but the API Docs guidelines (referenced as also applying to the other Guides in the Guides Guidelines) seemed to strongly favour that: https://guides.rubyonrails.org/api_documentation_guidelines.html#fixed-width-font

Reply:
That is true for the API docs, which were traditionally written in RDoc syntax, slowly being converted to markdown.

For guides, we just write markdown: https://guides.rubyonrails.org/ruby_on_rails_guides_guidelines.html